src/rules/deserializationOfUntrustedData.ts

```diff
@@ -0,0 +1,78 @@
+import { Event, EventNavigator } from '@appland/models';
+import { MatchResult, Rule, RuleLogic } from '../types';
+import { URL } from 'url';
+
+function sanitizesData(event: Event, objectId: number, label: string): boolean {
+  return (
+    event.labels.has(label) &&
+    !!event.returnValue &&
+    !!event.returnValue.object_id &&
+    event.returnValue.object_id === objectId
+  );
+}
+
+function* precedingEvents(rootEvent: Event, target: Event) {
+  for (const event of new EventNavigator(rootEvent).descendants()) {
+    if (event.event === target) {
+      break;
+    }
+    yield event;
+  }
+}
+
+function allArgumentsSanitized(rootEvent: Event, event: Event): boolean {
+  return (event.parameters || [])
+    .filter((parameter) => parameter.object_id)
+    .every((parameter): boolean => {
+      for (const candidate of precedingEvents(rootEvent, event)) {
+        if (sanitizesData(candidate.event, parameter.object_id!, Sanitize)) {
+          return true;
+        }
+      }
+      return false;
+    });
+}
+
+function build(): RuleLogic {
+  function matcher(rootEvent: Event): MatchResult[] | undefined {
+    for (const event of new EventNavigator(rootEvent).descendants()) {
+      if (
+        event.event.labels.has(DeserializeUnsafe) &&
+        !event.event.ancestors().find((ancestor) => ancestor.labels.has(DeserializeSafe))
+      ) {
+        if (allArgumentsSanitized(rootEvent, event.event)) {
+          return;
+        } else {
+          return [
+            {
+              level: 'error',
+              event: event.event,
+              message: `${event.event} deserializes untrusted data`,
```

Comment:
Protip: use destructuring to avoid noise. For example:
```suggestion
    for (const {event} of new EventNavigator(rootEvent).descendants()) {
      if (
        event.labels.has(DeserializeUnsafe) &&
        !event.ancestors().find(({labels}) => labels.has(DeserializeSafe))
      ) {
        if (allArgumentsSanitized(rootEvent, event)) {
          return;
        } else {
          return [
            {
              level: 'error',
              event,
              message: `${event} deserializes untrusted data`,
```